File: examples/conference/schema/src/documents/session.ts
import { defineField, defineType } from 'sanity';
import { PiPresentation as Icon } from 'react-icons/pi';

export const session = defineType({
  name: 'conf.session',
  title: 'Session',
  type: 'document',
  icon: Icon,
  fields: [
    defineField({
      name: 'edition',
      title: 'Edition',
      type: 'reference',
      to: [{ type: 'conf.edition' }],
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'title',
      title: 'Title',
      type: 'string',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'slug',
      title: 'Slug',
      type: 'slug',
      validation: (Rule) => Rule.required(),
      options: {
        source: 'title',
        maxLength: 64,
      },
    }),
    defineField({
      name: 'description',
      title: 'Description',
      type: 'text',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'speakers',
      title: 'Speakers',
      type: 'array',
      of: [{ type: 'reference', to: [{ type: 'conf.speaker' }] }],
    }),
    defineField({
      name: 'tracks',
      title: 'Tracks',
      type: 'array',
      of: [{ type: 'reference', to: [{ type: 'conf.track' }] }],
    }),
    defineField({
      name: 'duration',
      title: 'Duration (in minutes)',
      type: 'number',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'format',
      title: 'Format',
      type: 'string',
      validation: (Rule) => Rule.required(),
      initialValue: 'talk',
      options: {
        list: [
          { title: 'Intro', value: 'intro' },
          { title: 'Keynote', value: 'keynote' },
          { title: 'Talk', value: 'talk' },
          { title: 'Workshop', value: 'workshop' },
          { title: 'Panel', value: 'panel' },
          { title: 'Lightning Talk', value: 'lightning' },
          { title: 'Breakout', value: 'breakout' },
          { title: 'Networking', value: 'networking' },
          { title: 'Outro', value: 'outro' },
        ],
      },
    }),
    defineField({
      name: 'level',
      title: 'Level',
      type: 'string',
      validation: (Rule) => Rule.required(),
      initialValue: 'all',
      options: {
        list: [
          { title: 'Beginner', value: 'beginner' },
          { title: 'Intermediate', value: 'intermediate' },
          { title: 'Advanced', value: 'advanced' },
          { title: 'All Levels', value: 'all' },
        ],
      },
    }),
  ],
  preview: {
    select: {
      title: 'title',
      speakers: 'speakers',
      track: 'tracks[0].name',
      duration: 'duration',
    },
    prepare({ title, speakers, track, duration }) {
      const speakerCount = speakers?.length || 0;
      return {
        title,
        subtitle: `${track ? `${track} - ` : ''} ${speakerCount} speaker${speakerCount === 1 ? '' : 's'} (${duration} min)`,
      };
    },
  },
});

File: examples/conference/feature_conference/lib/content/session.dart
import 'package:feature_conference/layouts/session_layout.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:vyuh_core/vyuh_core.dart';
import 'package:vyuh_extension_content/vyuh_extension_content.dart';
import 'package:vyuh_feature_system/content/index.dart';

import '../content/speaker.dart';
import '../content/track.dart';

part 'session.g.dart';

enum SessionFormat {
  intro,
  keynote,
  talk,
  workshop,
  panel,
  lightning,
  breakout,
  networking,
  outro,
}

enum SessionLevel {
  beginner,
  intermediate,
  advanced,
  all,
}

@JsonSerializable()
class Session extends ContentItem {
  static const schemaName = 'conf.session';

  static final typeDescriptor = TypeDescriptor(
    schemaType: schemaName,
    fromJson: Session.fromJson,
    title: 'Session',
  );

  static final contentBuilder = ContentBuilder(
    content: typeDescriptor,
    defaultLayout: SessionLayout(),
    defaultLayoutDescriptor: SessionLayout.typeDescriptor,
  );

  static final descriptor = ContentDescriptor.createDefault(
    schemaType: schemaName,
    title: 'Session',
  );

  @JsonKey(name: '_id')
  final String id;

  final String title;
  final String? slug;
  final PortableTextContent? description;
  final int duration;
  final SessionFormat format;
  final SessionLevel level;
  final ObjectReference? edition;

  final List<Speaker>? speakers;

  final List<Track>? tracks;

  Session({
    required this.id,
    required this.title,
    this.slug,
    this.description,
    required this.duration,
    this.format = SessionFormat.talk,
    this.level = SessionLevel.all,
    this.edition,
    this.speakers,
    this.tracks,
    super.layout,
    super.modifiers,
  }) : super(schemaType: schemaName);

  factory Session.fromJson(Map<String, dynamic> json) =>
      _$SessionFromJson(json);
}
